src/bootstrap/src/core/build_steps/compile.rs

```diff
@@ -206,11 +222,19 @@ impl Step for Std {
             }
         }
 
-        let mut cargo = builder.cargo(compiler, Mode::Std, SourceType::InTree, target, "build");
-        std_cargo(builder, target, compiler.stage, &mut cargo);
-        for krate in &*self.crates {
-            cargo.arg("-p").arg(krate);
-        }
+        let mut cargo = if self.is_for_mir_opt_tests {
+            let mut cargo = builder.cargo(compiler, Mode::Std, SourceType::InTree, target, "rustc");
+            cargo.arg("-p").arg("std").arg("--crate-type=lib");
+            std_cargo(builder, target, compiler.stage, &mut cargo);
+            cargo
+        } else {
+            let mut cargo = builder.cargo(compiler, Mode::Std, SourceType::InTree, target, "build");
+            std_cargo(builder, target, compiler.stage, &mut cargo);
+            for krate in &*self.crates {
```

Comment:
So `self.crates` is entirely ignored when `is_for_mir_opt_tests` is set?

Reply:
Yes; as far as I understand `crates` is the crates you're asking to build explicitly. For example when you run `x build library/core` `crates` is `["core"]`. When we're building (possibly cross-compiling) a sysroot for mir-opt tests, the only thing it makes sense to build is `std` and all of its dependencies; attempting to select any other set would be unnecessary or insufficient to build some of the mir-opt tests.

Reply:
What about the "proc_macro" crate?

Reply:
That's handled down here: https://github.com/rust-lang/rust/pull/119035/files#diff-675065335a90de4a4e1ac5bc0a078709f865ff2d357ba2650309f0ad8614b1d3R1620-R1621

Reply:
And somehow it also works for Miri... :shrug: whatever, as long as CI is happy.^^ I just hope I won't have to debug this, the bootstrap logic is getting more and more messy it feels like.

Reply:
Miri doesn't use this code path, the only change for Miri is how we detect that we should disable the target sanity check.

The logic here is complicated mostly because the mir-opt tests cannot reuse Miri's technique for building a MIR-only sysroot because of this call: https://github.com/rust-lang/rust/blob/090d5eac722000906cc00d991f2bf052b0e388c3/compiler/rustc_mir_transform/src/inline/cycle.rs#L76, a check build doesn't produce MIR, a check build with -Zalways-encode-mir produces sporadic changes to inlining cyclic calls, and a normal build needs a linker to produce `libstd.so`.

I agree that the code is rather complicated, but I think that reflects our overall poor support for cross compilation as opposed to being new complexity. And I'm volunteering to help maintain this sort of code; I understand it now that I've worked on it a bit.